package org.freekode.tp2intervals.integration.platform.trainerroad.workout

import config.mock.ObjectMapperFactory
import config.mock.TRApiClientMock
import org.freekode.tp2intervals.domain.ExternalData
import org.freekode.tp2intervals.domain.TrainingType
import org.freekode.tp2intervals.domain.workout.structure.SingleStep
import org.freekode.tp2intervals.domain.workout.structure.WorkoutStructure
import org.freekode.tp2intervals.integration.platform.trainerroad.TRApiClientService
import org.freekode.tp2intervals.integration.platform.trainerroad.configuration.TrainerRoadConfiguration
import org.freekode.tp2intervals.integration.platform.trainerroad.configuration.TrainerRoadConfigurationRepository
import org.freekode.tp2intervals.integration.platform.trainerroad.member.TRUsernameRepository
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import org.mockito.Mockito.mock
import org.mockito.Mockito.`when`
import org.springframework.util.ResourceUtils

class TrainerRoadWorkoutRepositoryTest {
    private val objectMapper = ObjectMapperFactory.objectMapper()

    private val trainerRoadApiClient = TRApiClientMock(
        objectMapper,
        ResourceUtils.getFile("classpath:tr-workoutsdetails-simple.json").inputStream(),
        ResourceUtils.getFile("classpath:tr-workoutsdetails-complex.json").inputStream(),
        ResourceUtils.getFile("classpath:tr-workoutsdetails-another.json").inputStream(),
    )

    private val trainerRoadConfigurationRepository = trainerRoadConfigurationRepository()

    private val trApiClientService =
        TRApiClientService(trainerRoadApiClient, trainerRoadConfigurationRepository)

    private val TRWorkoutRepository =
        TRWorkoutRepository(mock(TRUsernameRepository::class.java), trApiClientService)

    @Test
    fun `should parse simple workout`() {
        // when
        val data = ExternalData(null, null, "simple")
        val workout = TRWorkoutRepository.getWorkoutFromLibrary(data)

        // then
        val structure = workout.structure!!

        assertEquals(TrainingType.VIRTUAL_BIKE, workout.details.type)
        assertEquals(WorkoutStructure.TargetUnit.FTP_PERCENTAGE, structure.target)
        assertEquals(11, structure.steps.size)
        assertEquals(5 * 60.toLong(), (structure.steps[0] as SingleStep).length.value)
        assertEquals(50, (structure.steps[0] as SingleStep).target.start)
        assertEquals(50, (structure.steps[0] as SingleStep).target.end)
        assertEquals(3 * 60.toLong(), (structure.steps[1] as SingleStep).length.value)
        assertEquals(60, (structure.steps[1] as SingleStep).target.start)
        assertEquals(60, (structure.steps[1] as SingleStep).target.end)
        assertEquals(13 * 60.toLong(), (structure.steps[2] as SingleStep).length.value)
        assertEquals(50, (structure.steps[2] as SingleStep).target.start)
        assertEquals(50, (structure.steps[2] as SingleStep).target.end)
    }

    @Test
    fun `should parse complex workout`() {
        // when
        val data = ExternalData(null, null, "complex")
        val workout = TRWorkoutRepository.getWorkoutFromLibrary(data)

        // then
        val structure = workout.structure!!

        assertEquals(TrainingType.VIRTUAL_BIKE, workout.details.type)
        assertEquals(WorkoutStructure.TargetUnit.FTP_PERCENTAGE, structure.target)
        assertEquals(23, structure.steps.size)
        assertEquals(4 * 60.toLong(), (structure.steps[0] as SingleStep).length.value)
        assertEquals(50, (structure.steps[0] as SingleStep).target.start)
        assertEquals(50, (structure.steps[0] as SingleStep).target.end)
        assertEquals(2 * 60.toLong(), (structure.steps[1] as SingleStep).length.value)
        assertEquals(55, (structure.steps[1] as SingleStep).target.start)
        assertEquals(55, (structure.steps[1] as SingleStep).target.end)
        assertEquals(2 * 60.toLong(), (structure.steps[2] as SingleStep).length.value)
        assertEquals(60, (structure.steps[2] as SingleStep).target.start)
        assertEquals(60, (structure.steps[2] as SingleStep).target.end)
    }

    @Test
    fun `should print wrong rest api response`() {
        // when
        val data = ExternalData(null, null, "another")
        val workout = TRWorkoutRepository.getWorkoutFromLibrary(data)

        // then
        val structure = workout.structure!!
        assertEquals(WorkoutStructure.TargetUnit.FTP_PERCENTAGE, structure.target)
        assertTrue(structure.steps.isNotEmpty())
    }

    @Test
    fun `should exclude html tags from description`() {
        val data = ExternalData(null, null, "simple")
        val workout = TRWorkoutRepository.getWorkoutFromLibrary(data)
        assertEquals(
            "simple is 4x3-minute intervals of leg-speed drills at a very low 60% FTP with 3 minutes of rest between intervals. " +
                    "Keep the pressure on the pedals light and your intensity low to moderate regardless of your cadence.",
            workout.details.description
        )
    }

    private fun trainerRoadConfigurationRepository(): TrainerRoadConfigurationRepository {
        val mock = mock(TrainerRoadConfigurationRepository::class.java)
        `when`(mock.getConfiguration()).thenReturn(TrainerRoadConfiguration(null, true))
        return mock
    }
}
